pyccel/decorators.py

```diff
@@ -38,6 +39,12 @@ def identity(f):
         return f
     return identity
 
+def template(name, types=[]):
+    """template decorator."""
+    def identity(f):
+        return f
+    return identity
```

Comment:
Do you not also need to pass the function `f`? You can test this by running your test code in pure python mode

Reply:
We don't actually need to decorate the function, passing the function `f` doesn't serve any purpose.

Reply:
Hmmm, this is an interesting question that probably deserves some discussion.

On the one hand, if the template can be used for multiple functions then it does not make sense for it to be a decorator

On the other hand, if we want to translate a function rather than an entire module we need the template to be attached to the function so that we have the information when we inspect the code

Reply:
Do you have a suggestion for what we could use instead? Maybe it should be an argument to the epyccel function to allow for reuse of the templates without code duplication?

Reply:
I don't think we have such problem, since we gather all the templates pragmas/decorators in `FunctionDef`, it's the samething we do for types, basically `template` and `types` works in the same manner.
what I meant before is that we don't need the actual python decoration to happen, we just need to be able to retrieve this information from the `ast`, this is why I favored not passing the function `f`, because it requires more checks and does the same thing as this.

Reply:
> I don't think we have such problem

Which problem exactly are you referring to?

> basically template and types works in the same manner

Yes, but `types` is always specific to a function

> what I meant before is that we don't need the actual python decoration to happen

You are right, apparently the code above does not cause an error 

> we just need to be able to retrieve this information from the ast

I would still double check that you can do that when you `epyccel`ize a single function rather than an entire module. You won't be able to write a test for this until issue #512 is fixed, but it would be worth checking that the case passes the semantic step. When we do:
```
import mod
f = epyccel(mod.test_func)
```
we do not have access to the entire code of `mod`. Instead we just have access to the specified function and decorators that are attached to it. If you don't pass the function then I wonder if the decorator is attached

Reply:
I am not sure, i thought since something like this works:
```
#$ header function f2(int)
#$ header function f1(int)
def f2(a):
    pass
def f1(a):
    pass
```
then this will work:
```
#$ header template T(int|real)
#$ header template U(int|real)
#$ header function f(T, U)
def f(a, b):
   pass
```
because I handled them in the same way. I'm going to test if `f = epyccel(mod.f)` passes the semantics as you suggested.
As for the decorators I presumed if the user decorated a function with a template then he wanted it to be used only for that function, so it's not available for other functions. In the other hand any template pragma written in the scope of the function or it's parents scopes are available to that function. 

> If you don't pass the function then I wonder if the decorator is attached

I checked the `ast` for this, the information is preserved.

Reply:
Actually epyccel doesn't work with header comments. Originally pyccel only used header comments, the decorators were added later in order to allow the user to epyccelize a function. So:
```
#$ header function f2(int)
#$ header function f1(int)
def f2(a):
    pass
def f1(a):
    pass
```
and
```
#$ header template T(int|real)
#$ header template U(int|real)
#$ header function f(T, U)
def f(a, b):
   pass
```
are both fine, but mean you have to use `pyccel` or epyccelize the whole module. This is the difference between decorators and header comments

> As for the decorators I presumed if the user decorated a function with a template then he wanted it to be used only for that function, so it's not available for other functions. 

That makes sense to me

> I checked the ast for this, the information is preserved.
Ok, then we should be fine